Lib/asyncio/graph.py

```diff
@@ -0,0 +1,247 @@
+"""Introspection utils for tasks call graphs."""
+
+import dataclasses
+import sys
+import types
+import typing
+
+from . import events
+from . import futures
+from . import tasks
+
+__all__ = (
+    'capture_call_graph',
+    'format_call_graph',
+    'print_call_graph',
+    'FrameCallGraphEntry',
+    'FutureCallGraph',
+)
+
+# Sadly, we can't re-use the traceback's module datastructures as those
+# are tailored for error reporting, whereas we need to represent an
+# async call graph.
+#
+# Going with pretty verbose names as we'd like to export them to the
+# top level asyncio namespace, and want to avoid future name clashes.
+
+
+@dataclasses.dataclass(frozen=True, slots=True)
+class FrameCallGraphEntry:
+    frame: types.FrameType
+
+
+@dataclasses.dataclass(frozen=True, slots=True)
+class FutureCallGraph:
+    future: futures.Future
+    call_stack: list[FrameCallGraphEntry]
+    awaited_by: list[FutureCallGraph]
+
+
+def _build_graph_for_future(future: futures.Future) -> FutureCallGraph:
+    if not isinstance(future, futures.Future):
+        raise TypeError(
+            f"{future!r} object does not appear to be compatible "
+            f"with asyncio.Future"
+        )
+
+    coro = None
+    if get_coro := getattr(future, 'get_coro', None):
+        coro = get_coro()
+
+    st: list[FrameCallGraphEntry] = []
+    awaited_by: list[FutureCallGraph] = []
+
+    while coro is not None:
+        if hasattr(coro, 'cr_await'):
+            # A native coroutine or duck-type compatible iterator
+            st.append(FrameCallGraphEntry(coro.cr_frame))
+            coro = coro.cr_await
+        elif hasattr(coro, 'ag_await'):
+            # A native async generator or duck-type compatible iterator
+            st.append(FrameCallGraphEntry(coro.cr_frame))
+            coro = coro.ag_await
+        else:
+            break
+
+    if future._asyncio_awaited_by:
+        for parent in future._asyncio_awaited_by:
+            awaited_by.append(_build_graph_for_future(parent))
+
+    st.reverse()
+    return FutureCallGraph(future, st, awaited_by)
+
+
+def capture_call_graph(
+    *,
+    future: futures.Future | None = None,
+    depth: int = 1,
+) -> FutureCallGraph | None:
```

Comment:
Functions from `traceback` module accept `limit=None` to limit the stack depth.
E.g.
>    Print up to *limit* stack trace entries (starting from the invocation
   point) if *limit* is positive.  Otherwise, print the last `abs(limit)`
   entries.  If *limit* is omitted or `None`, all entries are printed.

Maybe `graph.py` could follow the same schema for the sake of uniformness?

Reply:
Yeah, we probably should do that. @ambv your thoughts?

Reply:
This is a relatively minor change, I'll do that.